src/main/java/communication/MessageIDEvaluator.java

```diff
@@ -0,0 +1,142 @@
+package communication;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.log4j.Logger;
+
+import control.FBase;
+import de.hasenburg.fbase.model.GetMissedMessageResponse;
+import exceptions.FBaseCommunicationException;
+import exceptions.FBaseStorageConnectorException;
+import model.data.MessageID;
+import model.data.NodeID;
+
+/**
+ * The {@link MessageIDEvaluator} manages all messageIDs received.
+ * 
+ * Everytime,
+ * {@link Subscriber#interpreteReceivedEnvelope(model.messages.Envelope, org.zeromq.ZMQ.Socket)}
+ * is called, the subscriber adds the ID of the contained envelope if one is set using
+ * {@link #addReceivedMessageID(MessageID)}.
+ * 
+ * The {@link MessageIDEvaluator} continuously runs through all not evaluated messageIDs. If
+ * he finds a gap, he will create a messageSender and ask the related node about the send
+ * data.
+ * 
+ * @author jonathanhasenburg
+ *
+ */
+public class MessageIDEvaluator {
+
+	private static Logger logger = Logger.getLogger(MessageIDEvaluator.class.getName());
+
+	private FBase fBase;
+	private ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
+
+	private TreeMap<NodeID, TreeMap<String, TreeSet<Integer>>> idStorage = new TreeMap<>();
```

Comment:
what's the purpose of having a String machine ID here? I thought, messages were IDed by node.

Reply:
I thought that it might be better to have the message history size per machine (and not per node). That's why there is also the specific field in the config below. A reason for that is that each machine is responsible for a certain set of keygroups, and making the history size machine specific enables us to keep the last send messages for every keygroup, and not just the ones that are most used by all machines of the node. What do you think?